### PR TITLE
Fix mania GenerateHitResults() bugs

### PR DESCRIPTION
**Pull request by Syriiin**
There are a couple bugs:

1. An odd number of `Perfect`/`Great` hits would miss a hit and throw off following calculations.
2. `Ok`s were duplicated as `Meh`s

This function is quite hard to parse so it's not surprising these slipped in.
I think these variables could do with some renaming, but that's beyond the scope of this fix.

On a side note, my gut says it should just assume all `Great`s are `Perfect`s rather than splitting them down the middle.
Usually one would expect that omitting a parameter would set it as the best possible value.

### CODE REVIEW

File: PerformanceCalculator/Simulate/ManiaSimulateCommand.cs [hunk 1]
```diff
@@ -80,11 +80,12 @@ public class ManiaSimulateCommand : SimulateCommand
             int delta = targetTotal - remainingHits;
 
             // Each great and perfect increases total by 5 (great-meh=5)
-            // There is no difference in accuracy between them, so just halve arbitrarily.
-            greats = Math.Min(delta / 5, remainingHits) / 2;
-            int perfects = greats.Value;
+            // There is no difference in accuracy between them, so just halve arbitrarily (favouring perfects for an odd number).
+            int greatsAndPerfects = Math.Min(delta / 5, remainingHits);
+            greats = greatsAndPerfects / 2;
+            int perfects = greats.Value + (greatsAndPerfects % 2);
```
Comment (bdach) [inline PerformanceCalculator/Simulate/ManiaSimulateCommand.cs:86]: I would have done

```suggestion
            int perfects = greatsAndPerfects - greats.Value;
```

to dispel any and all doubts of inaccuracy. I guess what you have works but feels needlessly contrived?

**Comment by bdach** [comment]
> On a side note, my gut says it should just assume all `Great`s are `Perfect`s rather than splitting them down the middle. Usually one would expect that omitting a parameter would set it as the best possible value.

The reason I did this originally is that hitting all perfects seemed - excluding the easiest maps - rather unrealistic. Usually when someone isn't providing exact difficulty numbers they're providing raw accuracy so extrapolations need to be made.

And yeah that wouldn't matter one bit in score V1 but it does actually matter a bit now that perfects are scored 305 rather than 300 when it comes to accuracy at least.

File: PerformanceCalculator/Simulate/ManiaSimulateCommand.cs (hunk 1 above)
Comment (Syriiin) [reply PerformanceCalculator/Simulate/ManiaSimulateCommand.cs:86]: Personally I think divmod is a little more clear since the fact that it's integer division isn't immediately obvious.
But simply subtracting from that value is certainly more readable given c# doesn't have a built in divmod function.
I'll change it tonight :ok_hand:

**Comment by Syriiin** [comment]
> The reason I did this originally is that hitting all perfects seemed - excluding the easiest maps - rather unrealistic. Usually when someone isn't providing exact difficulty numbers they're providing raw accuracy so extrapolations need to be made.

Yea in practice for real plays it makes sense, it's just the expectation I think most people would have from difficulty calculators though.
To get the max pp from a score, you just give the beatmap with no parameters. :man_shrugging:
Needing to explicitly pass a `--greats 0` (especially when it's divergent from existing patterns) just feels wrong.
If someone explicitly passed accuracy as `--accuracy 99` or something, I can see splitting the difference between greats and perfects as more justifiable there.
Anyway, off topic!

**Review by bdach — APPROVED** [review]
